### PR TITLE
Use the idiomatic RAII lock releaser.

### LINKED ISSUE #106: Deadlock'd indexUpdateActor caused the unique index building to hang
After adding the tests in PR #105, a bug was exposed (yay), which happens when first insert some large docs and then try to add unique compound indexes. The symptom is that the index build will seemingly hang and client will eventually timeout. The hang was caused by the mysterious cancel actor message sent to `scanAndBuildIndex` actor.

The normal case is when a duplicate was detected and the unique index building is about to fail, it will first throw out the duplicated entry error, which was finally propagated to user, then the actor aforementioned got the actor canceled message.

In the bug, the duplicated entry error was, somehow, swallowed somewhere and thus client never got the error returned and thus thought the server hangs.

### PR DESCRIPTION
**Pull request by dongxinEric**
Now instead of manually releasing the lock and adding try-catch blocks across the code, simply use the idiomatic RAII style lock releaser whose dtor is guaranteed to be called once the ACTOR gets out of scope, i.e. died

This resolves #106

### CODE REVIEW

**Comment by dongxinEric** [comment]
Let me run some correctness tests before merge @apkar

File: src/QLContext.actor.cpp [hunk 1]
```diff
@@ -376,12 +375,13 @@ struct CompoundIndexPlugin : IndexPlugin, ReferenceCounted<CompoundIndexPlugin>,
...
-					// one finished duplication detecting and index record writing. And thus the following section
-					// before the `lock.release()` call, needs to be protected using a mutex lock.
-					Void _ = wait(self->flowControlLock.get()->lock->take(1));
-				}
+
+				// When building the unique index, each record out of the table scan needs to wait for the previous
+				// one finished duplication detecting and index record writing. And thus the following section
+				// before the `lock.release()` call, needs to be protected using a mutex lock.
+				state FlowLock::Releaser releaser(*self->flowControlLock.get(), 1);
+				Void _ = wait(self->flowControlLock.get()->take(1));
```
Comment (apkar) [inline src/QLContext.actor.cpp]: It's still `Optional`, why we are not checking for `present()`?

Comment (dongxinEric) [reply src/QLContext.actor.cpp]: Oh the `ASSERT` was lost through copy and paste. Let me add it back.

Comment (apkar) [reply src/QLContext.actor.cpp]: Ah! its because unique index precondition.

**Comment by dongxinEric** [comment]
@apkar  Please review this again since I made some changes. Also please do not enable the unique indexes in correctness test just yet, I noticed some inconsistent behavior between MM and DocLayer and will need to dig into that a bit.

File: src/QLContext.actor.cpp (hunk 1 above)
Comment (dongxinEric) [reply src/QLContext.actor.cpp]: Yeah the unique flag should have ensured it, but will add an ASSERT just to be safe.

**Comment by apkar** [comment]
@dongxinEric Did you run correctness?

**Comment by dongxinEric** [comment]
@dongxinEric yes and I did run the code with unique index enabled and found a bunch of discrepancies between MM result and DocLayer result. But nothing more than that. Let me run some with unique index disabled and will circle back with results.

**Comment by apkar** [comment]
My concern is not introducing new issues with this patch. Are the correctness issues related to this patch?

**Comment by dongxinEric** [comment]
Run 51K iterations and no new errors. @apkar

**Comment by dongxinEric** [comment]
@apkar The issues found are not related to the change.

**Pull request merged**